### PR TITLE
[WIP] Broadcast support in codegen

### PR DESCRIPTION
**Pull request by csarofeen**
- [x] Add braodcast type to iterDomain
- [x] Add broadcast IR node
- [x] Add broadcast op to Arith
- [ ] Work broadcast through transform ops and transform replay. 
- [ ] Work broadcast through rfactor
- [x] Figure out if we can support merging broadcast dims with non-broadcast dims
- [x] Add broadcast op to lowering logic: lower2gpu, lower loops, and unroll_loops
- [x] Add broadcast logic to index compute and predicate compute
- [ ] Test pointwise and broadcast operations
- [ ] Test broadcast operations before and after reduction
- [ ] Test softmax
- [ ] Test Normalization

### CODE REVIEW

File: aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h [hunk 1]
```diff
@@ -84,7 +84,10 @@ namespace at { namespace cuda {
   _(nvrtcCompileProgram)                         \
   _(cuCtxGetCurrent)                             \
   _(cuModuleUnload)                              \
-  _(cuDevicePrimaryCtxGetState)
+  _(cuDevicePrimaryCtxGetState)                  \
```
Comment (jjsjann123) [inline aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h]: Nope, we don't want this for rocm. It will break rocm build, as their driver API doesn't provide these.

This is something that we fixed in last PR.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 2]
```diff
@@ -33,33 +31,84 @@ TORCH_CUDA_API Val* newValLike(const Val* const val, DataType dtype) {
...
+  ValType out_vtype = vals[0]->getValType().value();
+  DataType out_dtype = vals[0]->getDataType().value();
+
+  for (auto val : vals) {
+    TORCH_CHECK(val->isVal(), "Invalid statement found during promotion.");
+    TORCH_CHECK(
+        val->getDataType().value() != DataType::Null,
+        "Invalid datatype found during prmotion.");
+    out_vtype = promote_type(out_vtype, val->getValType().value());
+    out_dtype = promote_type(out_dtype, val->getDataType().value());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:97]: This is the place where I was asking earlier.
out_dtype promotion should be weighted by ValType of the input.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 3]
```diff
@@ -238,6 +321,48 @@ TORCH_CUDA_API Val* sum(Val* v1, const std::vector<int>& axes) {
...
+        "Expected identity op, but didn't get a TensorView back.");
+    return static_cast<TensorView*>(identity);
+  }
+
+  std::vector<IterDomain*> out_domain;
+  unsigned int iinp = 0, ibdim = 0;
+  while (ibdim < is_broadcast_dim.size()) {
+    if (is_broadcast_dim[ibdim]) {
+      out_domain.push_back(new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:358]: note to myself: broadcast is supported, I need to update integration to remove the conversion to implicit broadcast (stride-0-with-broadcasted-shape).

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 4]
```diff
@@ -69,12 +69,12 @@ TensorDomain* IndexCompute::runBackward(std::vector<Expr*> history) {
   return running_td;
 }
 
-IndexCompute::IndexCompute(TensorDomain* td, std::vector<Val*> _indices)
+IndexCompute::IndexCompute(TensorDomain* td, const std::vector<Val*>& _indices)
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: We are using copy-then-move. Shouldn't use const-reference here.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp [hunk 5]
```diff
@@ -53,7 +53,7 @@ void IterVisitor::traverseFrom(
   FusionGuard fg(fusion);
   std::unordered_set<Statement*> visited;
   stmt_stack.clear();
-  stmt_stack.emplace_back(from.rbegin(), from.rend());
+  stmt_stack.emplace_back(std::vector<Statement*>(from.rbegin(), from.rend()));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: no we don't need `std::vector<Statement*>` here.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp [hunk 6]
```diff
@@ -64,7 +64,8 @@ void IterVisitor::traverseFrom(
 
     // Traverse down until we get to a leaf
     while (!next_stmts.empty()) {
-      stmt_stack.emplace_back(next_stmts.rbegin(), next_stmts.rend());
+      stmt_stack.emplace_back(
+          std::vector<Statement*>(next_stmts.rbegin(), next_stmts.rend()));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: same here.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp [hunk 7]
```diff
@@ -152,7 +153,7 @@ struct DependencyChains : public IterVisitor {
           deps.push_back(static_cast<Val*>(stack.back()));
       }
       // Order as dependency -> of
-      dep_chains.emplace_back(deps.rbegin(), deps.rend());
+      dep_chains.emplace_back(std::deque<Val*>(deps.rbegin(), deps.rend()));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: same here.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp [hunk 8]
```diff
@@ -169,7 +170,9 @@ struct DependencyChains : public IterVisitor {
       traverse(_dependency->fusion(), false);
   }
 
-  DependencyChains(std::set<Val*> _dependencies, bool all_chains_ = false)
+  DependencyChains(
+      const std::set<Val*>& _dependencies,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: This will break clang-tidy.
We should revert the changes.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 9]
```diff
@@ -562,11 +620,19 @@ bool Merge::sameAs(const Merge* const other) const {
 Reorder::Reorder(
     TensorDomain* _out,
     TensorDomain* _in,
-    std::vector<int> _new2old)
+    const std::vector<int>& _new2old)
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: revert the change here.

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 10]
```diff
@@ -236,8 +236,7 @@ struct Influence : public TransformIter {
 
   std::vector<bool> influence;
 
-  Influence(std::vector<bool> td_influence)
-      : influence(std::move(td_influence)) {}
+  Influence(const std::vector<bool>& td_influence) : influence(td_influence) {}
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp]: revert this change.

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 11]
```diff
@@ -515,7 +513,7 @@ struct Replay : public TransformIter {
   }
 
   std::vector<int> axis_map;
-  Replay(std::vector<int> _axis_map) : axis_map(std::move(_axis_map)) {}
+  Replay(const std::vector<int>& _axis_map) : axis_map(_axis_map) {}
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp]: revert this change.

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 12]
```diff
@@ -827,7 +828,7 @@ struct ReplaySelf : public TransformIter {
   }
 
   std::vector<int> axis_map;
-  ReplaySelf(std::vector<int> _axis_map) : axis_map(std::move(_axis_map)) {}
+  ReplaySelf(const std::vector<int>& _axis_map) : axis_map(_axis_map) {}
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp]: revert this change.

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 13]
```diff
@@ -1068,8 +1072,7 @@ struct TransformBackward : public TransformIter {
 
   std::vector<int> axis_map;
 
-  TransformBackward(std::vector<int> _axis_map)
-      : axis_map(std::move(_axis_map)){};
+  TransformBackward(const std::vector<int>& _axis_map) : axis_map(_axis_map){};
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp]: revert this change.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:97]: Please ignore the previous comment. This will require some further investigation to figure out the proper way to handle promotion. But it's not a blocking issue.

**Comment by jjsjann123** [comment]
LGTM except some minor merge conflicts that was handled in the wrong direction.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/arith.cpp:358]: Definitely not supported. Don't do that.

File: aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h (hunk 1 above)
Comment (csarofeen) [reply aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h]: Removed, thanks for catching that.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: yeah I think I put const ref separately from when you added move, my mistake.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Fixed.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Fixed.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp (hunk 8 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Ugh clang-tidy how I hate you.

**Comment by csarofeen** [comment]
Closing in favor of: https://github.com/csarofeen/pytorch/pull/53
